File: ChromaSchemes/core/role_binder.py
"""Bind generated colours to Prismata universal roles."""

from __future__ import annotations


UNIVERSAL_ROLES = [
    "primary_surface",
    "secondary_surface",
    "tertiary_surface",
    "accent_surface",
    "border_fg",
    "menu_fg",
    "menu_bg",
    "interactive_fg",
    "interactive_bg",
    "neutral_fg",
    "neutral_bg",
]


class RoleBinder:
    """Map colour IDs to semantic universal roles deterministically."""

    def bind(self, colour_set: list[dict]) -> dict:
        if not colour_set:
            raise ValueError("colour_set cannot be empty")

        refs = [entry["id"] for entry in colour_set]
        payload: dict[str, dict] = {}

        for index, role in enumerate(UNIVERSAL_ROLES):
            colour_ref = refs[index % len(refs)]
            channel = "fg" if role.endswith("_fg") else "bg"
            payload[role] = {"layers": {channel: {"colourRef": colour_ref}}}

        return payload

File: ChromaSchemes/tooling/roleBinder.ts
/**
 * RoleBinder – maps colour-set entries onto universal scheme roles.
 *
 * Description:
 *   Binds palette entries to the shared role vocabulary used by the preview
 *   and emitter stages.
 *
 * Inputs:
 *   - bind(colourSet)
 *     * colourSet: array of normalised colour-set entries.
 *
 * Outputs:
 *   A role map keyed by role name with foreground/background layer refs.
 *
 * Example:
 *   const binder = new RoleBinder();
 *   const roles = binder.bind(colourSet);
 */
import { IRoleBinder } from './interfaces';

const UNIVERSAL_ROLES = [
  'primary_surface',
  'secondary_surface',
  'tertiary_surface',
  'accent_surface',
  'border_fg',
  'menu_fg',
  'menu_bg',
  'interactive_fg',
  'interactive_bg',
  'neutral_fg',
  'neutral_bg',
];

export class RoleBinder implements IRoleBinder {
  bind(colourSet: Array<Record<string, unknown>>) {
    return UNIVERSAL_ROLES.reduce<Record<string, unknown>>((acc, role, idx) => {
      const ref = String(colourSet[idx % colourSet.length]?.id ?? 'c0');
      const layer = role.endsWith('_fg') ? 'fg' : 'bg';
      acc[role] = { layers: { [layer]: { colourRef: ref } } };
      return acc;
    }, {});
  }
}
